src/traits/wf.md

```diff
@@ -1,11 +1,440 @@
+// proving `Implemented(T: Iterator)`, and we are unable to prove that for an
+// unknown `T`.
+//
+// Hence, this type definition is considered illegal. An additional
+// `where T: Iterator` would make it legal.
+```
+
+# Trait definitions
+
+Given a general trait definition:
+```rust,ignore
+trait Trait<P1...> where WC_trait {
+    type Assoc<P2...>: Bounds_assoc where WC_assoc;
+}
+```
+
+we generate the following goal:
+```text
+forall<P1...> {
+    if (FromEnv(WC_trait)) {
+        WellFormed(InputTypes(WC_trait)) &&
+
+            forall<P2...> {
+                if (FromEnv(WC_assoc)) {
+                    WellFormed(InputTypes(Bounds_assoc)) &&
+                        WellFormed(InputTypes(WC_assoc))
+                }
+            }
+    }
+}
+```
+
+There is not much to verify in a trait definition. We just want
+to prove that the types appearing in the trait definition are well-formed,
+under the assumption that the different where clauses hold.
+
+Some examples:
+```rust,ignore
+trait Foo<T> where T: Iterator, <T as Iterator>::Item: Debug {
+    ...
+}
+// The only non-parameter type which appears in this definition is
+// `<T as Iterator>::Item`. The generated goal is the following:
+// ```
+// forall<T> {
+//     if (FromEnv(T: Iterator), FromEnv(<T as Iterator>::Item: Debug)) {
+//         WellFormed(<T as Iterator>::Item)
+//     }
+// }
+// ```
+// which is provable thanks to the `FromEnv(T: Iterator)` assumption.
+```
+
+```rust,ignore
+trait Bar {
+    type Assoc<T>: From<<T as Iterator>::Item>;
+}
+// The only non-parameter type which appears in this definition is
+// `<T as Iterator>::Item`. The generated goal is the following:
+// ```
+// forall<T> {
+//     WellFormed(<T as Iterator>::Item)
+// }
+// ```
+// which is not provable, hence the trait definition is considered illegal.
+```
+
+```rust,ignore
+trait Baz {
+    type Assoc<T>: From<<T as Iterator>::Item> where T: Iterator;
+}
+// The generated goal is now:
+// ```
+// forall<T> {
+//     if (FromEnv(T: Iterator)) {
+//         WellFormed(<T as Iterator>::Item)
+//     }
+// }
+// ```
+// which is now provable.
+```
+
+# Impls
+
+Now we give ourselves a general impl for the trait defined above:
+```rust,ignore
+impl<P1...> Trait<A1...> for SomeType<A2...> where WC_impl {
+    type Assoc<P2...> = SomeValue<A3...> where WC_assoc;
+}
+```
+
+Note that here, `WC_assoc` are the same where clauses as those defined on the
+associated type definition in the trait declaration, *except* that type
+parameters from the trait are substituted with values provided by the impl
+(see example below). You cannot add new where clauses. You may omit to write
+the where clauses if you want to emphasize the fact that you are actually not
+relying on them.
+
+Some examples to illustrate that:
+```rust,ignore
+trait Foo<T> {
+    type Assoc where T: Clone;
+}
+
+struct OnlyClone<T: Clone> { ... }
+
+impl<T> Foo<Option<T>> for () {
```

Comment:
Perhaps make the type parameter `U` instead of `T` for this impl, so that when you quote the bound `T: Clone`, is unambiguous with `Option<U>: Clone`.